File: examples/go/apps/minipi/scorer.go
// SPDX-License-Identifier: AGPL-3.0-or-later
// ScorerActor — trajectory scoring for eval pipelines.
//
// Demonstrates: LLM-as-judge pattern, heuristic scoring, rubric evaluation.
package main

import (
	"encoding/json"
	"fmt"

	"github.com/bhatti/PlexSpaces/sdks/go/plexspaces"
)

// ScorerActor scores agent trajectories against rubrics.
//
// Supports two scoring modes:
// - heuristic: rule-based scoring (fast, deterministic)
// - llm_judge: LLM-as-judge (expensive, flexible) — uses LLMGateway
type ScorerActor struct {
	plexspaces.BaseActor
	ActorID      string `json:"actor_id"`
	TotalScored  int    `json:"total_scored"`
}

func NewScorerActor() plexspaces.Actor {
	a := &ScorerActor{}
	a.SetSelf(a)
	return a
}

func (s *ScorerActor) Init(configJSON string) string {
	var config struct {
		ActorID string `json:"actor_id"`
	}
	if err := json.Unmarshal([]byte(configJSON), &config); err != nil {
		return "ERROR: " + err.Error()
	}
	s.SetRuntimeMetadata(config.ActorID)
	s.ActorID = config.ActorID
	if err := host.PG().Join("svc:scorer"); err != nil {
		host.Warn(fmt.Sprintf("ScorerActor: failed to join svc:scorer: %v", err))
	}
	host.Info(fmt.Sprintf("ScorerActor Init actor_id=%s", config.ActorID))
	return ""
}

func (s *ScorerActor) Handle(fromActor, msgType, payloadJSON string) string {
	p := parsePayload(payloadJSON)
	op := stringVal(p, "op", msgType)
	switch op {
	case "score":
		return s.score(p)
	case "batch_score":
		return s.batchScore(p)
	case "get_stats":
		return marshal(map[string]any{"status": "ok", "total_scored": s.TotalScored})
	default:
		return marshal(map[string]any{"error": "unknown_op", "op": op})
	}
}

func (s *ScorerActor) score(p map[string]any) string {
	traj, _ := p["trajectory"].(map[string]any)
	if traj == nil {
		return marshal(map[string]any{"error": "trajectory is required", "score": 0.0})
	}
	// rubric can be a string or a map
	rubricType := "task_completion"
	var rubric map[string]any
	switch rv := p["rubric"].(type) {
	case string:
		rubricType = rv
		rubric = map[string]any{"type": rv}
	case map[string]any:
		rubric = rv
		rubricType = stringVal(rv, "type", "task_completion")
	}
	if rubric == nil {
		rubric = map[string]any{"type": "task_completion"}
	}

	var sc float64
	var detail string
	switch rubricType {
	case "task_completion":
		sc, detail = s.scoreTaskCompletion(traj, rubric)
	case "tool_use":
		sc, detail = s.scoreToolUse(traj, rubric)
	case "efficiency":
		sc, detail = s.scoreEfficiency(traj, rubric)
	case "llm_judge":
		sc, detail = s.scoreLLMJudge(traj, rubric)
	default:
		sc, detail = s.scoreTaskCompletion(traj, rubric)
	}

	s.TotalScored++
	s.IncrCounter(host, "trajectories_scored")

	return marshal(map[string]any{
		"status":        "ok",
		"trajectory_id": stringVal(traj, "trajectory_id", ""),
		"score":         roundFloat(sc, 3),
		"rubric_type":   rubricType,
		"detail":        detail,
	})
}

func (s *ScorerActor) batchScore(p map[string]any) string {
	trajsRaw := sliceVal(p, "trajectories")
	rubric, _ := p["rubric"].(map[string]any)

	results := make([]any, 0, len(trajsRaw))
	scores := make([]float64, 0, len(trajsRaw))
	for _, t := range trajsRaw {
		tMap, _ := t.(map[string]any)
		if tMap == nil {
			continue
		}
		resp := parsePayload(s.score(map[string]any{
			"trajectory": tMap,
			"rubric":     rubric,
		}))
		results = append(results, resp)
		scores = append(scores, float64Val(resp, "score", 0.0))
	}

	meanScore := 0.0
	passRate := 0.0
	if len(scores) > 0 {
		sum := 0.0
		passed := 0
		for _, sc := range scores {
			sum += sc
			if sc >= 0.8 {
				passed++
			}
		}
		meanScore = sum / float64(len(scores))
		passRate = float64(passed) / float64(len(scores))
	}

	return marshal(map[string]any{
		"status":     "ok",
		"scores":     results,
		"mean_score": roundFloat(meanScore, 3),
		"pass_rate":  roundFloat(passRate, 3),
	})
}

func (s *ScorerActor) scoreTaskCompletion(traj, rubric map[string]any) (float64, string) {
	outcome := stringVal(traj, "outcome", "")
	steps := sliceVal(traj, "steps")
	expectedKeywordsRaw := sliceVal(rubric, "expected_keywords")
	expectedKeywords := extractStringSliceFromAny(expectedKeywordsRaw)

	baseScore := 0.1
	switch outcome {
	case "success", "completed":
		baseScore = 0.7
	case "budget_exceeded":
		baseScore = 0.3
	case "suspended":
		baseScore = 0.5
	}

	// Bonus for completing in fewer steps
	maxSteps := intVal(rubric, "max_steps", 20)
	stepCount := len(steps)
	if stepCount > 0 && stepCount <= maxSteps/2 {
		baseScore = minFloat(1.0, baseScore+0.15)
	}

	// Check for expected keywords in outputs
	allOutputs, _ := json.Marshal(steps)
	allOutputsStr := string(allOutputs)
	keywordMatches := 0
	for _, kw := range expectedKeywords {
		if containsAny(allOutputsStr, kw) {
			keywordMatches++
		}
	}
	if len(expectedKeywords) > 0 {
		keywordBonus := 0.15 * float64(keywordMatches) / float64(len(expectedKeywords))
		baseScore = minFloat(1.0, baseScore+keywordBonus)
	}

	detail := fmt.Sprintf("outcome=%s steps=%d keywords_matched=%d/%d", outcome, stepCount, keywordMatches, len(expectedKeywords))
	return baseScore, detail
}

func (s *ScorerActor) scoreToolUse(traj, rubric map[string]any) (float64, string) {
	steps := sliceVal(traj, "steps")
	expectedToolsRaw := sliceVal(rubric, "expected_tools")
	expectedTools := extractStringSliceFromAny(expectedToolsRaw)

	toolCalls := 0
	usedTools := map[string]bool{}
	for _, step := range steps {
		sm, _ := step.(map[string]any)
		if sm == nil {
			continue
		}
		if stringVal(sm, "kind", "") == "tool_call" {
			toolCalls++
			toolName := stringVal(sm, "tool_name", "")
			if toolName != "" {
				usedTools[toolName] = true
			}
		}
	}

	var score float64
	if len(expectedTools) == 0 {
		if toolCalls > 0 {
			score = 0.8
		} else {
			score = 0.4
		}
	} else {
		matches := 0
		for _, et := range expectedTools {
			if usedTools[et] {
				matches++
			}
		}
		score = float64(matches) / float64(len(expectedTools))
	}

	usedList := make([]string, 0, len(usedTools))
	for k := range usedTools {
		usedList = append(usedList, k)
	}
	detail := fmt.Sprintf("tool_calls=%d used_tools=%v expected=%v", toolCalls, usedList, expectedTools)
	return score, detail
}

func (s *ScorerActor) scoreEfficiency(traj, rubric map[string]any) (float64, string) {
	totalTokens := intVal(traj, "total_input_tokens", 0) + intVal(traj, "total_output_tokens", 0)
	budget := intVal(rubric, "token_budget", 4096)

	if totalTokens == 0 {
		return 0.5, "no token data"
	}
	efficiency := maxFloat(0.0, 1.0-float64(totalTokens)/float64(budget))
	outcome := stringVal(traj, "outcome", "")
	if outcome != "success" && outcome != "completed" {
		efficiency *= 0.5
	}
	detail := fmt.Sprintf("tokens=%d budget=%d outcome=%s", totalTokens, budget, outcome)
	return roundFloat(efficiency, 3), detail
}

func (s *ScorerActor) scoreLLMJudge(traj, rubric map[string]any) (float64, string) {
	llmID, err := registryFirst("llm_gateway", "svc:llm_gateway", "completion")
	if err != nil || llmID == "" {
		return s.scoreTaskCompletion(traj, rubric)
	}

	criteria := stringVal(rubric, "criteria", "Did the agent successfully complete the task?")
	trajSummary := map[string]any{
		"outcome":      stringVal(traj, "outcome", ""),
		"step_count":   len(sliceVal(traj, "steps")),
		"total_tokens": intVal(traj, "total_input_tokens", 0) + intVal(traj, "total_output_tokens", 0),
	}
	summaryJSON, _ := json.Marshal(trajSummary)

	prompt := fmt.Sprintf(`Rate this agent trajectory on a scale of 0.0 to 1.0.

Criteria: %s

Trajectory summary: %s

Respond with ONLY a JSON object: {"score": 0.0-1.0, "reasoning": "brief explanation"}`,
		criteria, string(summaryJSON))

	resp, askErr := askActor(llmID, "completion", map[string]any{
		"op":       "completion",
		"messages": []any{map[string]any{"role": "user", "content": prompt}},
	}, 15000)
	if askErr == nil && len(resp) > 0 {
		response, _ := resp["response"].(map[string]any)
		if response != nil {
			content := stringVal(response, "content", "")
			if content != "" {
				var parsed map[string]any
				if err2 := json.Unmarshal([]byte(content), &parsed); err2 == nil {
					return float64Val(parsed, "score", 0.5), stringVal(parsed, "reasoning", "")
				}
			}
		}
	}
	return s.scoreTaskCompletion(traj, rubric)
}

func minFloat(a, b float64) float64 {
	if a < b {
		return a
	}
	return b
}

func maxFloat(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}

File: examples/python/apps/minipi/scorer.py
# SPDX-License-Identifier: AGPL-3.0-or-later
"""ScorerActor — trajectory scoring for eval pipelines.

Demonstrates: LLM-as-judge pattern, heuristic scoring, rubric evaluation.
"""

import json
from plexspaces import actor, state, init_handler, handler, host


@actor
class ScorerActor:
    """
    Scores agent trajectories against rubrics.

    Supports two scoring modes:
    - heuristic: rule-based scoring (fast, deterministic)
    - llm_judge: LLM-as-judge (expensive, flexible) — uses LLMGateway
    """

    actor_id: str = state(default="")
    total_scored: int = state(default=0)

    @init_handler
    def on_init(self, config: dict) -> None:
        self.actor_id = config.get("actor_id", "")
        try:
            host.kv.put("svc:scorer", host.self_id())
        except Exception:
            pass
        try:
            host.registry.register(None, self.actor_id or host.self_id(), "actor", "",
                                   object_category="scorer")
        except Exception:
            pass
        host.info(f"ScorerActor init actor_id={self.actor_id}")

    @handler("score")
    def score(self, trajectory: dict = None, rubric = None) -> dict:
        """Score a trajectory against a rubric. Returns score 0.0–1.0."""
        if not trajectory:
            return {"error": "trajectory is required", "score": 0.0}
        if not rubric:
            rubric = {"type": "task_completion"}
        if isinstance(rubric, str):
            rubric = {"type": rubric}

        rubric_type = rubric.get("type", "task_completion")
        score = 0.0
        detail = ""

        if rubric_type == "task_completion":
            score, detail = self._score_task_completion(trajectory, rubric)
        elif rubric_type == "tool_use":
            score, detail = self._score_tool_use(trajectory, rubric)
        elif rubric_type == "efficiency":
            score, detail = self._score_efficiency(trajectory, rubric)
        elif rubric_type == "llm_judge":
            score, detail = self._score_llm_judge(trajectory, rubric)
        else:
            score, detail = self._score_task_completion(trajectory, rubric)

        self.total_scored += 1
        host.incr_counter("trajectories_scored", 1)

        return {
            "status": "ok",
            "trajectory_id": trajectory.get("trajectory_id", ""),
            "score": round(score, 3),
            "rubric_type": rubric_type,
            "detail": detail,
        }

    @handler("batch_score")
    def batch_score(self, trajectories: list = None, rubric: dict = None) -> dict:
        """Score multiple trajectories against the same rubric."""
        if not trajectories:
            return {"error": "trajectories is required", "scores": []}
        results = [self.score(t, rubric) for t in trajectories]
        scores = [r.get("score", 0.0) for r in results]
        return {
            "status": "ok",
            "scores": results,
            "mean_score": sum(scores) / max(len(scores), 1),
            "pass_rate": sum(1 for s in scores if s >= 0.8) / max(len(scores), 1),
        }

    @handler("get_stats")
    def get_stats(self) -> dict:
        return {"status": "ok", "total_scored": self.total_scored}

    # ------------------------------------------------------------------

    def _score_task_completion(self, traj: dict, rubric: dict) -> tuple:
        """Score based on outcome and expected keywords."""
        outcome = traj.get("outcome", "")
        steps = traj.get("steps", [])
        expected_keywords = rubric.get("expected_keywords", [])

        if outcome in ("success", "completed"):
            base_score = 0.7
        elif outcome == "budget_exceeded":
            base_score = 0.3
        elif outcome == "suspended":
            base_score = 0.5
        else:
            base_score = 0.4

        # Bonus for completing in fewer steps
        max_steps = rubric.get("max_steps", 20)
        step_count = len(steps)
        if step_count <= max_steps // 2:
            base_score = min(1.0, base_score + 0.15)

        # Check for expected keywords in final output
        all_outputs = json.dumps([s.get("output", "") for s in steps])
        keyword_matches = sum(1 for kw in expected_keywords if kw.lower() in all_outputs.lower())
        if expected_keywords:
            keyword_bonus = 0.15 * (keyword_matches / len(expected_keywords))
            base_score = min(1.0, base_score + keyword_bonus)

        detail = f"outcome={outcome} steps={step_count} keywords_matched={keyword_matches}/{len(expected_keywords)}"
        return base_score, detail

    def _score_tool_use(self, traj: dict, rubric: dict) -> tuple:
        """Score based on correct tool usage."""
        steps = traj.get("steps", [])
        tool_calls = [s for s in steps if s.get("kind") == "tool_call"]
        expected_tools = rubric.get("expected_tools", [])

        used_tools = {s.get("method", "").replace("tool:", "") for s in tool_calls}
        if not expected_tools:
            score = 0.8 if tool_calls else 0.4
        else:
            matches = len(set(expected_tools) & used_tools)
            score = matches / len(expected_tools)

        detail = f"tool_calls={len(tool_calls)} used_tools={list(used_tools)} expected={expected_tools}"
        return score, detail

    def _score_efficiency(self, traj: dict, rubric: dict) -> tuple:
        """Score based on token efficiency (fewer tokens = better)."""
        total_tokens = traj.get("total_input_tokens", 0) + traj.get("total_output_tokens", 0)
        budget = rubric.get("token_budget", 4096)

        if total_tokens == 0:
            return 0.5, "no token data"

        # Score inversely proportional to token usage
        efficiency = max(0.0, 1.0 - (total_tokens / budget))
        outcome = traj.get("outcome", "")
        if outcome != "success":
            efficiency *= 0.5  # Penalize failed runs

        detail = f"tokens={total_tokens} budget={budget} outcome={outcome}"
        return round(efficiency, 3), detail

    def _score_llm_judge(self, traj: dict, rubric: dict) -> tuple:
        """Use LLM-as-judge for flexible scoring."""
        llm_id = self._find_service("llm_gateway")
        if not llm_id:
            # Fall back to heuristic
            return self._score_task_completion(traj, rubric)

        criteria = rubric.get("criteria", "Did the agent successfully complete the task?")
        traj_summary = {
            "outcome": traj.get("outcome"),
            "step_count": len(traj.get("steps", [])),
            "total_tokens": traj.get("total_input_tokens", 0) + traj.get("total_output_tokens", 0),
        }

        prompt = f"""Rate this agent trajectory on a scale of 0.0 to 1.0.

Criteria: {criteria}

Trajectory summary: {json.dumps(traj_summary)}

Respond with ONLY a JSON object: {{"score": 0.0-1.0, "reasoning": "brief explanation"}}"""

        try:
            resp = host.ask(llm_id, "completion", {
                "messages": [{"role": "user", "content": prompt}]
            }, 15000)

            if resp and "error" not in resp:
                content = resp.get("response", {}).get("content", "")
                parsed = json.loads(content)
                return parsed.get("score", 0.5), parsed.get("reasoning", "")
        except Exception:
            pass

        return self._score_task_completion(traj, rubric)

    def _find_service(self, service_type: str) -> str:
        """Discover service actor ID via object registry; falls back to peer ID on same node."""
        try:
            regs = host.registry.discover(None, object_category=service_type, limit=1)
            if regs:
                return regs[0]["object_id"]
        except Exception:
            pass
        idx = self.actor_id.find("//")
        if idx >= 0:
            return service_type + self.actor_id[idx:]
        return service_type
